### PR TITLE
Quality upper bound

### PR DESCRIPTION
**Pull request by seelabs**
Fix bug in qualityUpperBound:

* In and Out parameters were swapped when calculating the rate
* In and out qualities were not calculated correctly; use existing functions
  to get the qualities
* Added tests to check that theoretical quality matches actual computed quality

### CODE REVIEW

File: src/ripple/app/paths/impl/StrandFlow.h [hunk 1]
```diff
@@ -320,6 +320,24 @@ struct FlowResult
...
+    {
+        if (std::tie(stepQ, dir) = step->qualityUpperBound(v, dir); stepQ)
+            q = composed_quality(q, *stepQ);
+        else
+            return boost::none;
+    }
+    return q;
+};
+/// @endcond
+
```
Comment (seelabs) [inline src/ripple/app/paths/impl/StrandFlow.h]: A quick comment to explain moving this function: I have an upcoming PR that requires `qualityUpperBound` be be declared earlier in this file, so this PR moved it to reduce conflicts with the upcoming PR.

File: src/ripple/app/paths/impl/DirectStep.cpp [hunk 2]
```diff
@@ -835,18 +835,26 @@ DirectStepI<TDerived>::qualityUpperBound(ReadView const& v, DebtDirection& dir)
...
-    auto dstQIn =
-        static_cast<TDerived const*>(this)->quality(v, QualityDirection::in);
 
-    if (isLast_ && dstQIn > QUALITY_ONE)
-        dstQIn = QUALITY_ONE;
     Issue const iss{currency_, src_};
-    return Quality(getRate(STAmount(iss, srcQOut), STAmount(iss, dstQIn)));
+    // Be careful not to switch the parameters to `getRate`. The
+    // `getRate(offerOut, offerIn)` function is usually used for offers. It
+    // returns offerIn/offerOut. For a direct step, the rate is dstQOut/srcQIn
```
Comment (thejohnfreeman) [inline src/ripple/app/paths/impl/DirectStep.cpp]: Small but important mixup:

> For a direct step, the rate is ~dstQOut/srcQIn~ srcQOut/dstQIn

File: src/ripple/protocol/Quality.h [hunk 3]
```diff
@@ -276,6 +276,40 @@ class Quality
...
+    // relative distance is abs(a-b)/min(a,b)
+    friend double
+    relativeDistance(Quality const& q1, Quality const& q2)
+    {
+        assert(q1.m_value > 0 && q2.m_value > 0);
+
+        if (q1.m_value == q2.m_value)  // make expected common case fast
+            return 0;
+
+        auto const minV = std::min(q1.m_value, q2.m_value);
```
Comment (thejohnfreeman) [inline src/ripple/protocol/Quality.h]: How about `auto const [minV, maxV] = std::minmax(q1.m_value, q2.m_value)`?

File: src/ripple/protocol/Quality.h [hunk 4]
```diff
@@ -276,6 +276,40 @@ class Quality
...
+        auto const maxV = std::max(q1.m_value, q2.m_value);
+
+        auto mantissa = [](std::uint64_t rate) {
```
Comment (thejohnfreeman) [inline src/ripple/protocol/Quality.h]: What is going on here? `Quality` is a "custom floating point representation" according to its comment, but the schema is not given. Were the implementations of these `mantissa` and `exponent` functions reverse-engineered from the implementation of `round`? What if we made `mantissa` and `exponent` methods of `Quality` (they can even be private), and called them from here and `round`?

Separately, documenting the schema is a perfect opportunity to use ASCII art in a comment.

**Comment by scottschurr** [comment]
Without having looked hard at the change yet, it seems like these changes will effect payment calculations.  Do we need an amendment to control the switchover?

**Comment by seelabs** [comment]
@scottschurr These changes do not effect payments; they only effect the new offer crossing code (ref: https://github.com/seelabs/rippled/blob/qualityUpperBound/src/ripple/app/paths/impl/StrandFlow.h#L531-L538). As that code isn't active (and I'm not even sure the offer crossing would be effected by this bug) I don't believe this needs an amendment.

Note: I do intend to use this in some upcoming payment code.

File: src/ripple/app/paths/impl/DirectStep.cpp (hunk 2 above)
Comment (seelabs) [reply src/ripple/app/paths/impl/DirectStep.cpp]: Nice catch! Fixed in 593ef30bf `[fold] Fix type and use std::minmax`

File: src/ripple/protocol/Quality.h (hunk 3 above)
Comment (seelabs) [reply src/ripple/protocol/Quality.h]: Love it! Fixed in 593ef30bf `[fold] Fix type and use std::minmax`

File: src/test/app/TheoreticalQuality_test.cpp [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+            std::array<Account, numAccounts> accounts{alice, bob, carol, dan};
+            static_assert(
+                numAccounts == 4, "Path is only correct for four accounts");
+            path const accountsPath(accounts[1], accounts[2]);
+            env.fund(XRP(10000), alice, bob, carol, dan);
+            env.close();
+
+            // iterate through all pairs of accounts, randomly set the transfer
+            // rate, qIn, qOut, and if the account issues or redeems
+            for (std::size_t i = 0; i < numAccounts; ++i)
```
Comment (scottschurr) [inline src/test/app/TheoreticalQuality_test.cpp]: This `i` shadows the outer `i`.  Maybe `j` would be a better choice (assuming no one thinks it's a `Journal`?  No, `j` is used below...  `k` maybe?

At any rate, the shadowing really ought to be fixed.  It's easy for a reader to get confused between the inner and outer `i`.

File: src/test/app/TheoreticalQuality_test.cpp [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+        // clang-format on
+
+        int const numTestIterations = reqNumIterations.value_or(200);
+
+        constexpr std::uint32_t paymentAmount = 1;
+
+        Currency const eurCurrency = to_currency("EUR");
+        Currency const usdCurrency = to_currency("USD");
+
+        // Class to randommly set account transfer rates, qualities, and other
```
Comment (scottschurr) [inline src/test/app/TheoreticalQuality_test.cpp]: randommly -> randomly?

File: src/ripple/app/paths/impl/BookStep.cpp [hunk 7]
```diff
@@ -426,21 +426,22 @@ bool BookStep<TIn, TOut, TDerived>::equal (Step const& rhs) const
...
 
     // This can be simplified (and sped up) if directories are never empty.
     Sandbox sb(&v, tapNONE);
     BookTip bt(sb, book_);
     if (!bt.step(j_))
-        return boost::none;
+        return {boost::none, dir};
 
-    return static_cast<TDerived const*>(this)->qualityUpperBound(
+    auto const q = static_cast<TDerived const*>(this)->qualityUpperBound(
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/app/paths/impl/BookStep.cpp]: I agree, that API is confusing. I'll think about it and try to address it in the next patch.

File: src/ripple/app/paths/impl/DirectStep.cpp [hunk 8]
```diff
@@ -854,7 +853,8 @@ DirectStepI<TDerived>::qualityUpperBound(ReadView const& v, DebtDirection& dir)
     // (Input*dstQIn/srcQOut = Output; So rate = srcQOut/dstQIn). Although the
     // first parameters is called `offerOut`, it should take the `dstQIn`
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/app/paths/impl/DirectStep.cpp]: Fixed in 8ac0095f7 `[fold] Fix spelling errors and rename some variables`

File: src/ripple/app/paths/impl/DirectStep.cpp [hunk 9]
```diff
@@ -829,12 +829,11 @@ DirectStepI<TDerived>::lineQualityIn (ReadView const& v) const
 }
 
 template <class TDerived>
-boost::optional<Quality>
-DirectStepI<TDerived>::qualityUpperBound(ReadView const& v, DebtDirection& dir)
+std::pair<boost::optional<Quality>, DebtDirection>
+DirectStepI<TDerived>::qualityUpperBound(ReadView const& v, DebtDirection prevStepDebtDir)
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/app/paths/impl/DirectStep.cpp]: Fixed in 8ac0095f7 `[fold] Fix spelling errors and rename some variables`

File: src/test/app/TheoreticalQuality_test.cpp [hunk 10]
```diff
@@ -0,0 +1,568 @@
...
+            return;
+
+        // Due to the floating point calculations, theoretical and actual
+        // qualities are not expected to always be exactly equal. However, they
+        // should always be very close. This function checks that that two
+        // qualites are "close enough".
+        auto compareClose = [](Quality const& q1, Quality const& q2) {
+            // relative diff is fabs(a-b)/min(a,b)
+            // can't get access to internal value. Use the rate
+            double const tolerence = 0.0000001;
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/app/TheoreticalQuality_test.cpp]: Fixed in 8ac0095f7 `[fold] Fix spelling errors and rename some variables`

File: src/test/app/TheoreticalQuality_test.cpp [hunk 11]
```diff
@@ -0,0 +1,568 @@
...
+
+        // There are three relavent trust lines: `alice->bob`, `bob->carol`, and
+        // `carol->dan`. There are four accounts. If we count the number of
+        // combinations of parameters where a parameter is changed from its
+        // default value, there are
+        // 2^(num_trust_lines*num_trust_qualities+numAccounts) combinations of
+        // values to test, or 2^13 combinations. Use this value to set the
+        // number of iterations. Note however that many of these parameter
+        // combinations run essentially the same test. For example, changing the
+        // quality values for bob and carol are test almost the same thing.
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/app/TheoreticalQuality_test.cpp]: Fixed in 8ac0095f7 `[fold] Fix spelling errors and rename some variables`

File: src/test/app/TheoreticalQuality_test.cpp (hunk 5 above)
Comment (seelabs) [reply src/test/app/TheoreticalQuality_test.cpp]: Fixed in 8ac0095f7 `[fold] Fix spelling errors and rename some variables`

File: src/test/app/TheoreticalQuality_test.cpp (hunk 6 above)
Comment (seelabs) [reply src/test/app/TheoreticalQuality_test.cpp]: Fixed in 8ac0095f7 `[fold] Fix spelling errors and rename some variables`

File: src/ripple/app/paths/impl/DirectStep.cpp [hunk 12]
```diff
@@ -854,7 +853,8 @@ DirectStepI<TDerived>::qualityUpperBound(ReadView const& v, DebtDirection& dir)
     // (Input*dstQIn/srcQOut = Output; So rate = srcQOut/dstQIn). Although the
     // first parameters is called `offerOut`, it should take the `dstQIn`
     // variable.
-    return Quality(getRate(STAmount(iss, dstQIn), STAmount(iss, srcQOut)));
+    return {Quality(getRate(STAmount(iss, dstQIn), STAmount(iss, srcQOut))),
+            dir};
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/app/paths/impl/DirectStep.cpp]: This is how our current `.clang_format` indents this. I'm pretty agnostic about formatting. We can try to change the `.clang_format` rules if you don't like this (but I don't think this is the PR for it).

File: src/ripple/app/paths/impl/XRPEndpointStep.cpp [hunk 13]
```diff
@@ -241,15 +241,14 @@ inline bool operator==(XRPEndpointStep<TDerived> const& lhs,
...
-boost::optional<Quality>
+std::pair<boost::optional<Quality>, DebtDirection>
 XRPEndpointStep<TDerived>::qualityUpperBound(
-    ReadView const& v, DebtDirection& dir) const
+    ReadView const& v, DebtDirection prevStepDir) const
 {
-    dir = this->debtDirection(v, StrandDirection::forward);
-    return Quality{STAmount::uRateOne};
+    return {Quality{STAmount::uRateOne},
+            this->debtDirection(v, StrandDirection::forward)};
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/app/paths/impl/XRPEndpointStep.cpp]: See other comment, but TL;DR, this is how our current `.clang_format` indents this.

File: src/test/app/TheoreticalQuality_test.cpp [hunk 14]
```diff
@@ -0,0 +1,568 @@
...
+            RippleCalcTestParams rcp{env.json(
+                pay(accounts.front(), accounts.back(), iou(paymentAmount)),
+                accountsPath,
+                txflags(tfNoRippleDirect))};
+
+            testCase(rcp, env.closed());
+        }
+    }
+
+    void
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/app/TheoreticalQuality_test.cpp]: Agreed. I'll try to address this.

File: src/ripple/protocol/Quality.h (hunk 4 above)
Comment (seelabs) [reply src/ripple/protocol/Quality.h]: Agreed, the "custom floating point representation" isn't good documentation. We may as well try to address this in this PR. Will do.

File: src/ripple/app/paths/impl/BookStep.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/app/paths/impl/BookStep.cpp]: Fixed in 89f618124 `[fold] Rename an overload of qualityUpperBound to adjustQualityWithFees`. I renamed the overload and replaced the `auto` with an explicit type.

File: src/ripple/protocol/Quality.h (hunk 4 above)
Comment (seelabs) [reply src/ripple/protocol/Quality.h]: Added a comment `a1a67d6d0 Add fix amendment`. It's the same format as STAmount.

File: src/test/app/TheoreticalQuality_test.cpp [hunk 15]
```diff
@@ -0,0 +1,568 @@
...
+    // the balance will be zero. Since we are testing quality measures, the
+    // payment should not use multiple qualities, so the initialBalance
+    // needs to be able to handle an entire payment (otherwise an account
+    // will go from redeeming to issuing and the fees/qualities can change)
+    std::uint32_t const initialBalance_;
+
+    // probability of changing a value from its default
+    double const probChangeDefault_ = 0.75;
+    // probability that an account redeems into another account
+    double const probRedeem_ = 0.5;
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/app/TheoreticalQuality_test.cpp]: Fixed in `d43bd766a * [fold] Make some vars constexpr`

File: src/test/app/TheoreticalQuality_test.cpp [hunk 16]
```diff
@@ -0,0 +1,568 @@
...
+            relativeDistance(toQuality(100, 0), toQuality(100, 90)) == 1e90);
+        // Make the mantissa in the smaller value bigger than the mantissa in
+        // the larger value. Instead of checking the exact result, we check that
+        // it's large. If the values did not compare correctly in
+        // `relativeDistance`, then the returned value would be negative.
+        BEAST_EXPECT(
+            relativeDistance(toQuality(102, 0), toQuality(101, 90)) >= 1e89);
+    }
+
+    void
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/app/TheoreticalQuality_test.cpp]: I lowed the default number of iterations and added a comment as-to how to use the command line to set it. I agree, the test was taking a very long time to run (it still takes a long time, but it's more in line with other tests). Fixed in: `95aaa87d3 * qualityUpperBound [fold] Unit test iterations changes:
`

File: src/test/app/TheoreticalQuality_test.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/app/TheoreticalQuality_test.cpp]: I changed the unit tests to use a deterministic seed in this commit: `464b5cbda * [fold] Use a deterministic seed for unit test
`

**Comment by seelabs** [comment]
@scottschurr @thejohnfreeman Took care of the last few issues and re-based on the latest develop.

File: src/ripple/app/paths/impl/DirectStep.cpp [hunk 17]
```diff
@@ -153,8 +153,8 @@ class DirectStepI : public StepImp<IOUAmount, IOUAmount, DirectStepI<TDerived>>
     std::uint32_t
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/app/paths/impl/DirectStep.cpp]: Fixed in `4f655c917 * [fold] Add missing include, cleanup comments, & minor code cleanups:
`

File: src/ripple/app/paths/impl/DirectStep.cpp [hunk 18]
```diff
@@ -821,24 +821,50 @@ DirectStepI<TDerived>::lineQualityIn (ReadView const& v) const
...
 {
-    auto const prevStepDebtDir = dir;
-    dir = this->debtDirection(v, StrandDirection::forward);
-    std::uint32_t const srcQOut = [&]() -> std::uint32_t {
-        if (redeems(prevStepDebtDir) && issues(dir))
-            return transferRate(v, src_).value;
-        return QUALITY_ONE;
+    if (!v.rules().enabled(fixQualityUpperBound))
+    {
+        auto const dir = this->debtDirection(v, StrandDirection::forward);
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/app/paths/impl/DirectStep.cpp]: Fixed in `4f655c917 * [fold] Add missing include, cleanup comments, & minor code cleanups:
`

File: src/ripple/app/paths/impl/Steps.h [hunk 19]
```diff
@@ -176,18 +176,24 @@ class Step
...
-       @param dir in/out param. Set to DebtDirection::redeems if the previous step redeems.
-       Will be set to DebtDirection::redeems if this step redeems; Will be set to DebtDirection::issues if this
-       step does not redeem
-       @return The upper bound of quality for the step, or boost::none if the
-       step is dry.
+       @param prevStepDir Set to DebtDirection::redeems if the previous step redeems.
+       @return A pair. The first element is the upper bound of quality for the step, or boost::none if the
+       step is dry. The second element will be set to DebtDirection::redeems if this steps redeems,
+       DebtDirection:issues if this step issues.
+       @note it is an upper bound because offers on the books may be unfunded.
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/app/paths/impl/Steps.h]: Yeah, It's easy to forget why it's an upper-bound and not an actual quality.

File: src/ripple/protocol/Quality.h [hunk 20]
```diff
@@ -270,6 +274,39 @@ class Quality
...
+
+        auto const minVMantissa = mantissa(minV);
+        auto const maxVMantissa = mantissa(maxV);
+        auto const expDiff = exponent(maxV) - exponent(minV);
+
+        double const minVD = static_cast<double>(minVMantissa);
+        double const maxVD = expDiff ? maxVMantissa * pow(10, expDiff)
+                                     : static_cast<double>(maxVMantissa);
+
+        // maxVD and minVD are scaled so the have the same exponents. Dividing
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/ripple/protocol/Quality.h]: Fixed in `4f655c917 * [fold] Add missing include, cleanup comments, & minor code cleanups:
`

File: src/test/unity/app_test_unity2.cpp [hunk 21]
```diff
@@ -25,11 +25,12 @@
 #include <test/app/RCLCensorshipDetector_test.cpp>
 #include <test/app/RCLValidations_test.cpp>
 #include <test/app/Regression_test.cpp>
+#include <test/app/SHAMapStore_test.cpp>
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/unity/app_test_unity2.cpp]: Lol. Yeah, this is how clang-format does it, so let's keep that.

File: src/test/app/TheoreticalQuality_test.cpp [hunk 22]
```diff
@@ -0,0 +1,555 @@
...
+        maybeInsertQuality(jv, QualityDirection::out);
+        env(jv);
+        env.close();
+    };
+
+public:
+    explicit RandomAccountParams(
+        std::uint32_t trustAmount = 100,
+        std::uint32_t initialBalance = 50)
+        : engine_{1977u}  // Use a deterministic seed
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/app/TheoreticalQuality_test.cpp]: Agree 100% `use deterministic seed` is a pretty useless comment. Everyone can see it's a deterministic seed - the useful thing to know is why. Good feedback.

File: src/test/app/TheoreticalQuality_test.cpp [hunk 23]
```diff
@@ -0,0 +1,555 @@
...
+
+    void
+    maybeSetInitialBalance(
+        jtx::Env& env,
+        jtx::Account const& acc,
+        jtx::Account const& peer,
+        Currency const& currency)
+    {
+        using namespace jtx;
+        // Set the initial balance
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/app/TheoreticalQuality_test.cpp]: Lol.

Fixed in `4f655c917 * [fold] Add missing include, cleanup comments, & minor code cleanups:
`

File: src/test/app/TheoreticalQuality_test.cpp [hunk 24]
```diff
@@ -0,0 +1,555 @@
...
+            // relative diff is fabs(a-b)/min(a,b)
+            // can't get access to internal value. Use the rate
+            constexpr double tolerance = 0.0000001;
+            return relativeDistance(q1, q2) <= tolerance;
+        };
+
+        for (auto const& strand : sr.second)
+        {
+            Quality const theoreticalQ = *qualityUpperBound(sb, strand);
+            auto f = flow<IOUAmount, IOUAmount>(
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/app/TheoreticalQuality_test.cpp]: Fixed in `4f655c917 * [fold] Add missing include, cleanup comments, & minor code cleanups:
`

File: src/test/app/TheoreticalQuality_test.cpp [hunk 25]
```diff
@@ -0,0 +1,555 @@
...
+        constexpr std::uint32_t paymentAmount = 1;
+
+        Currency const eurCurrency = to_currency("EUR");
+        Currency const usdCurrency = to_currency("USD");
+
+        // Class to randomly set account transfer rates, qualities, and other
+        // params.
+        RandomAccountParams rndAccParams;
+
+        // Sped up tests by creating the environment outside the loop
```
[the remark this answers is not in the document]
Comment (seelabs) [reply src/test/app/TheoreticalQuality_test.cpp]: Fixed in `4f655c917 * [fold] Add missing include, cleanup comments, & minor code cleanups:
`

**Comment by seelabs** [comment]
@scottschurr Good feedback, thanks! I addressed them all in `4f655c917 * [fold] Add missing include, cleanup comments, & minor code cleanups:
`

**Comment by scottschurr** [comment]
FWIW, the clang-5 Travis build (only) is failing.  Here are the complaints:
```
In file included from ../../src/test/unity/app_test_unity2.cpp:33:
../../src/test/app/TheoreticalQuality_test.cpp:369:55: error: suggest braces around initialization of subobject [-Werror,-Wmissing-braces]
            std::array<Account, numAccounts> accounts{alice, bob, carol, dan};
                                                      ^~~~~~~~~~~~~~~~~~~~~~
                                                      {                     }
../../src/test/app/TheoreticalQuality_test.cpp:447:17: error: suggest braces around initialization of subobject [-Werror,-Wmissing-braces]
                alice, bob, carol, dan, oscar};
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                {                            }
2 errors generated.
```
Looks like we need double braces to initialize a couple of `std::array`s in order for clang-5 to be happy.

**Comment by seelabs** [comment]
Thanks @scottschurr I fixed clang 5 CI failures with in the `61aa2e8bd * qualityUpperBound [fold] Fix clang 5 warnings` patch.

**Comment by scottschurr** [comment]
Hmmm.  Clang-5 is good now.  Thanks @seelabs. Now the unit tests for the two debug Windows builds are failing.  Parallel unit test problems?  I'm not sure...

**Comment by seelabs** [comment]
@scottschurr The current beta has similar windows issues. I'll wait until those are resolved and see if they resolves these as well (I suspect they will)

File: src/ripple/app/paths/impl/DirectStep.cpp [hunk 26]
```diff
@@ -821,24 +823,49 @@ DirectStepI<TDerived>::lineQualityIn (ReadView const& v) const
...
+
+        if (isLast_ && dstQIn > QUALITY_ONE)
+            dstQIn = QUALITY_ONE;
+        Issue const iss{currency_, src_};
+        return {Quality(getRate(STAmount(iss, srcQOut), STAmount(iss, dstQIn))),
+                dir};
+    }
+
+    auto const [srcQOut, dstQIn] = [&] {
+        if (dir == DebtDirection::redeems)
```
Comment (thejohnfreeman) [inline src/ripple/app/paths/impl/DirectStep.cpp]: Other areas of the code use `redeems(dir)` (including 15 lines above). Do we want to use it everywhere? Possibly nowhere, i.e. remove it?

File: src/ripple/app/paths/impl/DirectStep.cpp [hunk 27]
```diff
@@ -821,24 +823,49 @@ DirectStepI<TDerived>::lineQualityIn (ReadView const& v) const
...
+        {
+            return qualitiesSrcRedeems(v);
+        }
+        else
+        {
+            return qualitiesSrcIssues(v, prevStepDir);
+        }
```
Comment (thejohnfreeman) [inline src/ripple/app/paths/impl/DirectStep.cpp]: Can this block become the following?

```cpp
auto const [srcQOut, dstQIn] = redeems(dir) ? qualitiesSrcRedeems(v) : qualitiesSrcIssues(v, prevStepDir);
```

Comment (seelabs) [reply src/ripple/app/paths/impl/DirectStep.cpp]: Yes, that's nicer. Fixed in `2847bf988 [fold] Replace lambda with conditional operator
`

File: src/ripple/app/paths/impl/DirectStep.cpp (hunk 26 above)
Comment (seelabs) [reply src/ripple/app/paths/impl/DirectStep.cpp]: I should have used the helper function here. Fixed in `2847bf988 [fold] Replace lambda with conditional operator `

As for removing the helper function: The `dir == DebtDirection::redeems` gets obnoxious quickly, but of course the helper function means people have to look up what that does at first. On balance, I think the code is better with it, but I'm OK removing it if you feel strongly. I do see the argument for removing it.

Comment (thejohnfreeman) [reply src/ripple/app/paths/impl/DirectStep.cpp]: I'm fine either way, was just listing options to surface our preferences.

**Comment by seelabs** [comment]
Squashed

**Pull request closed without merging**